roofit/roofit/src/RooLagrangianMorphFunc.cxx

```diff
@@ -0,0 +1,3522 @@
+               "morphing physics distributions between configurations. authors:"
+            << std::endl;
+  std::cout << "   "
+            << "Lydia Brenner   (lbrenner@cern.ch)" << std::endl;
+  std::cout << "   "
+            << "Carsten Burgard (cburgard@cern.ch)" << std::endl;
+  std::cout << "   "
+            << "Katharina Ecker (kecker@cern.ch)" << std::endl;
+  std::cout << "   "
+            << "Adam Kaluza     (akaluza@cern.ch)" << std::endl;
+  std::cout << "please feel free to contact with questions and suggestions."
+            << std::endl;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// calculate the number of samples needed to morph a bivertex, 2-2 physics
+/// process
+
+int RooLagrangianMorphFunc::countSamples(int nprod, int ndec, int nboth) {
+  FeynmanDiagram diagram;
+  std::vector<bool> prod;
+  std::vector<bool> dec;
+  for (int i = 0; i < nboth; ++i) {
+    prod.push_back(true);
+    dec.push_back(true);
+  }
+  for (int i = 0; i < nprod; ++i) {
+    prod.push_back(true);
+    dec.push_back(false);
+  }
+  for (int i = 0; i < ndec; ++i) {
+    prod.push_back(false);
+    dec.push_back(true);
+  }
+  diagram.push_back(prod);
+  diagram.push_back(dec);
+  MorphFuncPattern morphfuncpattern;
+  ::collectPolynomials(morphfuncpattern, diagram);
+  return morphfuncpattern.size();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// calculate the number of samples needed to morph a certain physics process
+/// usage:
+///   countSamples ( { RooLagrangianMorphFunc::makeHCggfCouplings(),
+///   RooLagrangianMorphFunc::makeHCHZZCouplings() } )
+
+int RooLagrangianMorphFunc::countSamples(std::vector<RooArgList *> &vertices) {
+  RooArgList operators, couplings;
+  for (auto vertex : vertices) {
+    extractOperators(*vertex, operators);
+    extractCouplings(*vertex, couplings);
+  }
+  FeynmanDiagram diagram;
+  ::fillFeynmanDiagram(diagram, vertices, couplings);
+  MorphFuncPattern morphfuncpattern;
+  ::collectPolynomials(morphfuncpattern, diagram);
+  return morphfuncpattern.size();
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// create TPair containers of the type expected by the RooLagrangianMorph
+
+TPair *RooLagrangianMorphFunc::makeCrosssectionContainer(double xs,
+                                                         double unc) {
+  TPair *v = new TPair(new TParameter<double>("xsection", xs),
+                       new TParameter<double>("uncertainty", unc));
+  return v;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// create only the weight formulas. static function for external usage.
+
+std::map<std::string, std::string> RooLagrangianMorphFunc::createWeightStrings(
+    const RooLagrangianMorphFunc::ParamMap &inputs,
+    const std::vector<std::string> &couplings) {
+  return RooLagrangianMorphFunc::createWeightStrings(inputs, couplings);
+}
+
```

Comment:
```suggestion
```
This function calls itself recursively in an infinite loop. Can we just remove this function? I doesn't bring any benefits except for that you can write
```C++
createWeightStrings(inputs, couplings);
```
instead of
```C++
createWeightStrings(inputs, {couplings});
```
which is not much longer.